### PR TITLE
feat!: Log Directive support for traces

### LINKED ISSUE #95: Switch PWG to return a list of logs
# Problem

Currently `Directive::Log` exclusively works with println statements. We directly write the information inside of `LogInfo` to std output. This is not ideal as there are multiple ways that someone way want to access logging information in their Noir circuit (writing to file, log traces in memory, etc.). 

The current design also requires us to pass down a flag during acir gen for determining whether to print logs during tests. More details can be found in this PR (https://github.com/noir-lang/noir/pull/630). 

# Solution

Having `Directive::Log` collect logs into a vector that is returned from the PWG would be the ideal solution. This would then enable whomever is calling `backend.solve` to handle the logs as they see fit. Such as printing to std out, writing to file, or accessing a vector of logs in memory. 

# Alternatives considered

There are not many alternatives if we want to enable devs to access the PWG logs in memory. Also it is a superior design to have whomever is calling `backend.solve` decide how to handle the system logs (nargo, JS packages). 

# Additional context

Relevant feature requested in Noir that requires this change: https://github.com/noir-lang/noir/issues/688

### PR DESCRIPTION
**Pull request by vezenovm**
# Related issue(s)

Resolves #95 

# Description

`Directive::Log` no longer prints the evaluated log while it is being evaluating during partial witness generation. This enables us to have both trace functionality and move logic for displaying logs to the code that calls the PWG rather than keeping it within the solver itself.

## Summary of changes

solve under the PWG now takes in a vector of `SolvedLog` structs. These solved logs contain either be a finalized string (calculated during compilation) or a vector of field elements that should be handled by the caller of `solve`. 

Several more changes showing how to handle logs correctly will be supplied in nargo (where we call `solve`)

## Dependency additions / changes

(If applicable.)

## Test additions / changes

As this reduces the functionality of the log directive, implementation of tests can be seen in the noir/nargo PR that implements these changes.

# Checklist

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [X] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [X] I have reviewed the changes on GitHub, line by line.
- [X] I have ensured all changes are covered in the description.

# Additional context

(If applicable.)

### CODE REVIEW

File: acir/src/circuit/directives.rs [hunk 1]
```diff
@@ -239,15 +247,26 @@ impl Directive {
...
-// evaluation, we can form an output string during ACIR generation.
-// Otherwise, we must store witnesses whose values will
-// be fetched during the PWG stage.
-pub enum LogInfo {
+/// This info is used when solving the initial witness
+/// If values are compile time and/or known during
+/// evaluation, we can form an output string during ACIR generation.
+/// Otherwise, we must store witnesses whose values will
+/// be fetched during the PWG stage.
+pub enum LogOutputInfo {
```
Comment (guipublic) [inline acir/src/circuit/directives.rs:255]: I don't understand why it is one or the other, why can't you have some of the witness solved during ACIR generation and some of them not solved?

File: acir/src/circuit/directives.rs [hunk 2]
```diff
@@ -239,15 +247,26 @@ impl Directive {
...
 }
 
+pub struct SolvedLog {
+    pub is_trace: bool,
+    pub output_info: SolvedLogOutputInfo,
+}
+
+pub enum SolvedLogOutputInfo {
+    FinalizedOutput(String),
+    WitnessValues(Vec<FieldElement>),
```
Comment (guipublic) [inline acir/src/circuit/directives.rs:267]: If the log is solved, why do you keep the values and not just have a string?

File: acvm/src/lib.rs [hunk 3]
```diff
@@ -74,7 +79,11 @@ pub trait PartialWitnessGenerator {
                         Self::solve_black_box_function_call(initial_witness, bb_func)
                     }
                     Opcode::Directive(directive) => {
-                        Self::solve_directives(initial_witness, directive)
+                        Self::solve_directives(initial_witness, directive).map(|possible_log| {
```
Comment (guipublic) [inline acvm/src/lib.rs:82]: I don't think the solve should take care of the logs, this should be handled internally by the log directive.

**Review by guipublic — CHANGES_REQUESTED** [review]
I think ACIR generation should produce only one log directive, so that we do not try to solve multiple times the same witness.
Then this solve directive would have a vector of 'field or witness', depending on whether the requested log was solved during acir generation.
PWG can completely ignore this directive, because the caller knowing the solved witness and the directive can figure out on its own.
This keeps ACVM simpler.

File: acir/src/circuit/directives.rs (hunk 1 above)
Comment (vezenovm) [reply acir/src/circuit/directives.rs:257]: There is certain situations where ACIR generation does not solve the witness. For example, when implementing the log directive for the first time it was discovered that attempting to print the output of blackbox functions required waiting until PWG to fetch the witness value.

File: acir/src/circuit/directives.rs (hunk 2 above)
Comment (vezenovm) [reply acir/src/circuit/directives.rs:269]: I didn't want to leave any decoding of the field elements. Transforming `Vec<FieldElement>` into a string right away would also require the extra logic that currently exists in the ACVM where we check whether we have a singular FieldElement or [multiple](https://github.com/noir-lang/acvm/blob/0b305b7bc18cc9f1569b4ddb09ac50a48ecef5f9/acvm/src/pwg/directives.rs#L172) (as the string looks different based on whether we have a singular field element or an array).

File: acvm/src/lib.rs (hunk 3 above)
Comment (vezenovm) [reply acvm/src/lib.rs:82]: I can move this logic to be internal to `solve_directives`

**Comment by vezenovm** [comment]
> I think ACIR generation should produce only one log directive, so that we do not try to solve multiple times the same witness.
> Then this solve directive would have a vector of 'field or witness', depending on whether the requested log was solved during acir generation.
> PWG can completely ignore this directive, because the caller knowing the solved witness and the directive can figure out on its own.
> This keeps ACVM simpler.

I see what you mean about how the caller knowing the solved witness and the directive can figure out what to call on its own.  This is something I hadn't considered and will look at making this change here and the upstream Noir PR. 

I have a couple clarifying questions:

When you refer to solve directive do you mean `Directive::Log` itself or the new `SolvedLog` type I made in this PR? `SolvedLog` isn't necessarily another directive just a new type that we return for solved logs. 

If we were to only have a vector of fields and witnesses, the log directive would then require some extra information such as whether the vector of fields/witnesses is a string or a normal array of fields. You can see how I currently handle this in the Noir PR when evaluating [array logs](https://github.com/noir-lang/noir/pull/961/files#diff-1c46fe8b1afaf36ca27510239b464f4a4b4969ce805e034a5d61b791c040e079R323). I foresee if I made this change that `Directive::Log` would have to have another field similar to `trace_label` called `is_string_output`. I think this is ok though as we already have to store extra information in order to implement traces and the caller would be the one handling any log directive flags. What do you think?

**Comment by vezenovm** [comment]
The VM has diverged a lot since this was opened, and these changes are small. Closing in favor of a fresh PR.

**Pull request closed without merging**